File: snippets/java/Snippet-001.java
import java.util.Objects;

class PriceCalculator {

    public class Bread implements Product {

        public static final float DISCOUNT_PERCENT = 0.1f;
        public static final int DISCOUNT_THRESHOLD = 10;
        public static final int NO_DISCOUNT = 0;

        public Float price() {
            return 3f;
        }

        public Float discount(Integer productQuantity) {
            return productQuantity >= DISCOUNT_THRESHOLD ? DISCOUNT_PERCENT : NO_DISCOUNT;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            return o != null && getClass() == o.getClass();
        }

        @Override
        public int hashCode() {
            return Objects.hash(this);
        }

        @Override
        public String toString() {
            return "Product{type=BREAD}";
        }
    }

    public class Candy implements Product {

        public static final int DISCOUNT_THRESHOLD = 100;
        public static final float DISCOUNT_PERCENT = 0.12f;
        public static final int NO_DISCOUNT = 0;

        public Float price() {
            return 0.49f;
        }

        public Float discount(Integer productQuantity) {
            return productQuantity >= DISCOUNT_THRESHOLD ? DISCOUNT_PERCENT : NO_DISCOUNT;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            return o != null && getClass() == o.getClass();
        }

        @Override
        public int hashCode() {
            return Objects.hash(this);
        }

        @Override
        public String toString() {
            return "Product{type=CANDY}";
        }
    }

    public interface Product {
        Float price();

        Float discount(Integer count);
    }
}

File: snippets/js/Snippet-001.js
const Type = {
    BREAD: 'BREAD', CANDY: 'CANDY'
}

class Product {
    constructor(type) {
        this.type = type;
    }
}


class PriceCalculator {
    price (product) {
        switch (product.type) {
            case Type.BREAD:  return 3;
            case Type.CANDY: return 0.49;
        }
    }

    discount (product, count) {
        switch (product.type) {
            case Type.BREAD: if (count >= 10) return 0.1; else return 0;
            case Type.CANDY: if (count >= 100) return 0.12; else return 0;
        }
    }
}
